resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -153,6 +156,25 @@
     final Future<Channel> channelFuture;
     final DatagramChannel ch;
 
+    // Comparator that ensures we will try first to use the nameservers that use our preferred address type.
+    private final Comparator<InetSocketAddress> nameServerComparator = new Comparator<InetSocketAddress>() {
+        @Override
+        public int compare(InetSocketAddress addr1, InetSocketAddress addr2) {
+            if (addr1.equals(addr2)) {
+                return 0;
+            }
+            if (!addr1.isUnresolved() &&
+                    preferredAddressType().addressType().isAssignableFrom(addr1.getAddress().getClass())) {
+                return 1;
+            }
+            if (!addr2.isUnresolved() &&
+                    preferredAddressType().addressType().isAssignableFrom(addr2.getAddress().getClass())) {
+                return -1;
+            }
+            return 1;
```

Comment:
IIUC this `Comparator` violates the interface:

```java
     * The implementor must ensure that <tt>sgn(compare(x, y)) ==
     * -sgn(compare(y, x))</tt> for all <tt>x</tt> and <tt>y</tt>.  (This
     * implies that <tt>compare(x, y)</tt> must throw an exception if and only
     * if <tt>compare(y, x)</tt> throws an exception.)<p>
```

example:
```java
InetSocketAddress addr1 = InetSocketAddress.createUnresolved("foo", 123);
        InetSocketAddress addr2 = InetSocketAddress.createUnresolved("bar", 456);
        int x = nameServerComparator.compare(addr1, addr2);
        int y = nameServerComparator.compare(addr2, addr1);
        assert x == -y;
```

there is also the transitive property to consider:

```
     * The implementor must also ensure that the relation is transitive:
     * <tt>((compare(x, y)&gt;0) &amp;&amp; (compare(y, z)&gt;0))</tt> implies
     * <tt>compare(x, z)&gt;0</tt>.<p>
```

Also if there are multiple addresses that are preferred & resolved the comparator may have issues.